src/lib.rs

```diff
@@ -0,0 +1,626 @@
+                        listener(Err(UStatus::fail_with_code(
+                            UCode::Internal,
+                            "Unable to get payload encoding",
+                        )));
+                        return;
+                    };
+                    UPayload {
+                        length: Some(0),
+                        format: encoding,
+                        data: Some(Data::Value(value.payload.contiguous().to_vec())),
+                    }
+                }
+                None => UPayload {
+                    length: Some(0),
+                    format: UPayloadFormat::UpayloadFormatUnspecified as i32,
+                    data: None,
+                },
+            };
+            // Create UMessage
+            let msg = UMessage {
+                source: Some(method.clone()),
+                attributes: Some(u_attribute.clone()),
+                payload: Some(u_payload),
+            };
+            if let Some(reqid) = u_attribute.reqid {
+                query_map.lock().unwrap().insert(reqid.into(), query);
+            } else {
+                listener(Err(UStatus::fail_with_code(
+                    UCode::Internal,
+                    "The request is without reqid in UAttributes",
+                )));
+                return;
+            }
+            listener(Ok(msg));
+        };
+        if let Ok(queryable) = self
+            .session
+            .declare_queryable(&zenoh_key)
+            .callback_mut(callback)
+            .res()
+            .await
+        {
+            self.queryable_map
+                .lock()
+                .unwrap()
+                .insert(hashmap_key.clone(), queryable);
+        } else {
+            return Err(UStatus::fail_with_code(
+                UCode::Internal,
+                "Unable to register callback with Zenoh",
+            ));
+        }
+
+        Ok(hashmap_key)
+    }
+    async fn unregister_rpc_listener(&self, method: UUri, listener: &str) -> Result<(), UStatus> {
+        // Do the validation
+        UriValidator::validate(&method)
+            .map_err(|_| UStatus::fail_with_code(UCode::InvalidArgument, "Invalid topic"))?;
+        // TODO: Check whether we still need method or not (Compare method with listener?)
+
+        if self
+            .queryable_map
+            .lock()
+            .unwrap()
+            .remove(listener)
+            .is_none()
+        {
+            return Err(UStatus::fail_with_code(
+                UCode::InvalidArgument,
+                "Listener doesn't exist",
+            ));
+        }
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl UTransport for UPClientZenoh {
+    async fn authenticate(&self, _entity: UEntity) -> Result<(), UStatus> {
+        // TODO: Not implemented
+        Err(UStatus::fail_with_code(
+            UCode::Unimplemented,
+            "Not implemented",
+        ))
+    }
+
+    async fn send(
```

Comment:
Hey @evshary :wave: 

There are some spec clarifications that will need to be made that have an impact on the implementation of send() and register_listener().

Because of how [uP Dispatchers work](https://github.com/eclipse-uprotocol/up-spec/tree/main/up-l2/dispatchers) over uP-L1 only, there's a need for a way to interact through up-client-zenoh-rust using only send() and register_listener(all_remote) that can intelligently beneath the hood operate with all of UMessageType::UMessageTypeRequest, UMessageType::UMessageTypeResponse, UMessageType::UMessageTypePublish.

I was looking at a couple of sequence diagrams last week with @stevenhartley that explained the rationale. @stevenhartley  -- could we open a PR to uc-spec with those in them?

Essentially the use case of uStreamer implies that it should call `UTransport::register_listener(special_mode_for_all_UAuthority, all_remote_callback)` and `all_remote_callback` should be triggered for all CEs whether UMessageType::UMessageTypeRequest, UMessageType::UMessageTypeResponse, or UMessageType::UMessageTypePublish

On the other end, since uStreamer talks over uP-L1 only, that means it should always use `UTransport::send(uuri, payload, attributes)` and the transport implementation in up-client-zenoh-rust should send appropriately whether UMessageType::UMessageTypeRequest, UMessageType::UMessageTypeResponse, or UMessageType::UMessageTypePublish. So in the Zenoh case where a pub/sub model is used for UMessageTypePublish and Queryables are used for UMessageTypeRequest & UMessageTypeResponse there'd have to be some thought given there on how to implement that.

If any questions / concerns we can discuss here or in this week's meeting.

Reply:
After viewing the flow chart mentioned in slack, I understand more now.
Let me put the diagram here for discussion.
![image](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/assets/456210/b8aae3c5-93d0-4f64-868d-41ab7cbc8ccb)

It's a little bit different from my initial thought.

1. While calling `invoke_method` on uPClientFoo side, I thought uStreamer will also call `invoke_method` on uPClientBar side, but it seems like we call the `send` directly to RpcListener on uPClientBar side.
2. While we call registerRpcListener (UUri: /uService/rpc.myMethod) on uPClientBar side, I thought we'll create the corresponding registerRpcListener (UUri: /uService/rpc.myMethod)  on uPClientFoo side to wait for `invoke_method` on uPClientFoo side. Then we don't need to listen to all the UUri (up/**).

To be honest, I don't have a strong preference on the uStreamer implementations, and UPClientZenoh should be able to support them no matter which implementation we choose.
It's great to know that I have a better understanding on the design now.

Reply:
Yup, I'll expand a little bit to make sure we're on the same page:

1. Correct, as we're bridging transports and don't want the implementation details of a given up-client-foo-lang to leak into the uStreamer, the uStreamer needs to be a little "dumb" and interact through these simpler uP-L1 APIs.
2. As you can see (think we need to update the diagram's uStreamer UUri), the uStreamer would be listening in on `up/*/**` to catch _any_ CEs originating externally from the Bar transport. The uStreamer would then catch that `UMessage`, and call `send()` on the up-client-foo-lang and under the hood, up-client-foo-lang would then do the appropriate action for the UMessageType (Request, Response, Publish).

Does that make sense?